06_puppy/alextmz/mapstore.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"fmt"
+)
+
+type MapStore map[int]Puppy
+
+func NewmapStore() MapStore {
+	a := MapStore{}
+	return a
+}
+
+func (m MapStore) CreatePuppy(p *Puppy) error {
+	switch {
+	case p == nil:
+		return fmt.Errorf("error, puppy pointer is nil")
+	case p.ID == 0:
+		p.ID = len(m) + 1
+		m[p.ID] = *p
+		return nil
+	default:
+		return fmt.Errorf("trying to create a puppy that already has an ID %d", p.ID)
+	}
+}
+
+func (m MapStore) ReadPuppy(id int) (Puppy, error) {
+	if id < 0 {
+		return Puppy{}, fmt.Errorf("invalid puppy id %d", id)
+	}
+	if v, ok := m[id]; ok {
+		return v, nil
```

Comment:
typically bail on error first, i.e. `!ok` ... `return err`

Reply:
Agree.
Was a bit puzzled by this one as it also affects my `switch` above, then I found https://golang.org/doc/effective_go.html#if, https://news.ycombinator.com/item?id=16678209, https://danp.net/2015/11/02/reducing-go-nesting.html. So despite not having an impact on readability of my code as there's no nesting it seems a good habit to have.